### DISCUSSION TITLE
Better understanding of alerting conditions & additional scrape configurations

### DISCUSSION THREAD

**Discussion by hartfordfive** [prometheus-operator/prometheus-operator#4837] [Q&A]
## Description
I'm looking at migrating some existing infrastructure with basic helm charts over to an operator-managed prometheus stack.   I would like to get a better understanding into the following topics:

### 1. Alerting conditions
I can see that alerting rules are contained within the [PrometheusRule](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md) resource although I can't seem to find out if the operator will automatically reload the rules configurations in the prometheus instances if they have an update (calling the `/-/reload` endpoint)?   I assume this must be the case although I wanted to confirm.  

Secondly, how are alerting conditions evaluated with the operator setup?  Traditionally, the alerting conditions are configured on the Prometheus shard instances (assuming funtional sharding with HA) that are scraping the metrics.   In this case, if sharding is done automatically, I assume this is why the ThanosRuler resource exists, which then users the sidecar to query the individual shards?  I know there are advantages and disadvantages to using thanos ruler instead of running the rules directly on a prometheus shard although what has been your experience with reliability and uptime?

### 2. Additional scrape configurations
I'm aware that scrape configurations are [defined within a secret](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/additional-scrape-config.md#creating-an-additional-configuration) and then referenced within the relevant Prometheus resource.   I also wanted to confirm that these are also automatically reloaded (calling the `/-/reload` endpoint) in Prometheus should their be an updated to the relevant secret containing the updated scrape configurations?

### 3.  ServiceMonitor/PodMonitor vs Prometheus Kubernetes native service discovery
I see the **ServiceMonitor** and **PodMonitor** resource can allow users to more easily expose metrics of their applications to Prometheus although I'd like to understand if there are any pros/cons comparted to use the native [kubernetes service discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config)?   What happens in  case where you need to collect metrics from applications on other kubernetes clusters?

### 4.  Degree of difficulting running prometheus-operator with thanos-operator

As I currently also use thanos (querier/store//compactor) for a global-view of metrics on all shards as well as long-term metrics retention, I'd like to get an idea of the degree of complexity to combine this operator with the thanos-operator.   Are there any special considerations or known issues, or should it be a relatively straight-forward process given users already have existing administration and configuration knowledge of thanos?

I appreciate any feedback!

**Comment by simonpasquier** [comment]
- Prometheus rules are reloaded automatically by the config reloader sidecar whenever a change is detected. Same goes for additional scrape configurations.
- Unless you have a strong need to use Thanos Ruler, I would stick with Prometheus only.
- The operator translates service and pod monitors into Prometheus Kubernetes service discovery configurations. The advantage with service/pod monitors is that you can manage everything as Kubernetes resources. As for collecting metrics from external Kubernetes clusters, this isn't something that we advise but additional scrap configs are here for that.
- I've got no experience with thanos-operator so I can't say much about it.

**Reply by hartfordfive** [reply]
@simonpasquier 

Regarding the execution of the alerting conditions and Thanos Ruler, I realize that it's always best to stick to having the evaluation of rules locally although I was trying to get a lower-level understanding  on how the operator evaluates alerting conditions when you have multiple shards.   

In a less elegant approach, you could implement functional sharding and have each prometheus shard scrapes it's own subset of metrics although the downside is that users need to know the correct shard which to add an alerting condition.  Not having yet studied how the operator is implemented and its resources, I assumed that it was using Thanos Ruler in the background to evaluate alerting conditions as we can't know ahead of time on which Prometheus shard to place a given alerting condition .  

Having done a bit more digging, I think I've now understood (please correct me if I'm wrong) that all alerting rules specified via the [PrometheusRule](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md#rule-selection) resource are combined then added to each prometheus shard.   I understand this model of running all alerting conditions follows the best practice (also described in [Thanos Ruler risks](https://thanos.io/tip/components/rule.md/#risk)) although this seems highly inefficient to run many alerting rules on all servers which may only have  subset of the metrics they're attempting to observe?   Also, I was concerned on the type of impact this may have in cases where queries use the ["**absent**" function](https://prometheus.io/docs/prometheus/latest/querying/functions/#absent), which would evaluate to true on X-1 shards (where X = number of shards).   I coincidentally came across this recent [github issue](https://github.com/prometheus-operator/kube-prometheus/issues/1744) which appears to describe this exact problem.   Can you confirm this is in fact an issue?   If so, what kind of solution could be proposed to remedy this other than running these queries via Thanos ruler?

On a related note, could the documentation be updated to provide a slightly more detailed explanation of how alerting rules are combined, deployed and executed at a lower level within the context of this operator?

Thanks again for your help!
